sdcm/remote.py

```diff
@@ -963,3 +972,463 @@ def run_output_check(self, cmd, timeout=None, ignore_status=False,
+    def __init__(self, hostname, user="root", port=22, password="",
+                 key_file=None, wait_key_installed=0, extra_ssh_options=""):
+
+        self.hostname = hostname
+        self.key_file = key_file
+        self.user = user
+        self.port = port
+        self.password = password
+        self.extra_ssh_options = extra_ssh_options
+        logger = logging.getLogger('avocado.test')
+        self.log = SDCMAdapter(logger, extra={'prefix': str(self)})
+
+        self.connect_timeout = 60
+        self._use_rsync = False
+        self.known_hosts_file = tempfile.mkstemp()[1]
+        config = Config(overrides={
+                        'load_ssh_config': False,
+                        'UserKnownHostsFile': self.known_hosts_file,
+                        'ServerAliveInterval': 300,
+                        'StrictHostKeyChecking': 'no'})
+        self.connection = Connection(
+            self.hostname,
+            self.user,
+            port=self.port,
+            config=config,
+            connect_timeout=self.connect_timeout,
+            connect_kwargs={
+                'key_filename': os.path.expanduser(self.key_file),
+            })
+        self._check_install_key_required()
+
+    def __str__(self):
+        return 'Remote [%s@%s]' % (self.user, self.hostname)
+
+    def ssh_debug_cmd(self):
+        if self.key_file:
+            return "SSH access -> 'ssh -i %s %s@%s'" % (self.key_file,
+                                                        self.user,
+                                                        self.hostname)
+        else:
+            return "SSH access -> 'ssh %s@%s'" % (self.user,
+                                                  self.hostname)
+
+    def run(self, cmd, timeout=None, ignore_status=False,
+            connect_timeout=300, options='', verbose=True,
+            args=None, log_file=None, watch_stdout_pattern=None, retry=0):
+
+        self.connection.connect_timeout = connect_timeout
+
+        for i in range(retry + 1):
+            try:
+                if verbose:
+                    self.log.debug('Run command {}'.format(cmd))
+
+                hide = not verbose
+                start_time = time.time()
+                if log_file:
+                    with open(log_file, 'a+') as fd_logfile:
+                        result = self.connection.run(cmd, warn=ignore_status,
+                                                     encoding='utf-8', hide=hide,
+                                                     out_stream=fd_logfile,
+                                                     err_stream=fd_logfile)
+                else:
+                    result = self.connection.run(cmd, warn=ignore_status, encoding='utf-8', hide=hide)
+
+                setattr(result, 'duration', time.time() - start_time)
+                setattr(result, 'exit_status', result.exited)
+            except (UnexpectedExit, Failure) as details:
+                self.log.error('{}'.format(details))
```

Comment:
error should be more informative 
`error executing command: {cmd}\ndetails: {details}`.format(**locals())

Reply:
Used the default str format of Fabric exception. But it caused another exception, if Invoke or paramiko exception happened. switch to your suggestions.